### ISSUE TITLE
Bidirectional Inference with Generic Callback Paramspec

### ISSUE THREAD

**Issue by hmc-cs-mdrissi** [microsoft/pyright#5147]
**Describe the bug**
```python
from typing import Callable, Generic, TypeVar

from typing_extensions import Concatenate, ParamSpec

T = TypeVar("T")
P = ParamSpec("P")


class Callback(Generic[T]):
    def __init__(self, func: Callable[Concatenate[T, P], object], *args: P.args, **kwargs: P.kwargs) -> None:
        ...


x: Callback[tuple[int, int]] = Callback(lambda p: (p[1], p[0]))  # Error
y = Callback[tuple[int, int]](lambda p: (p[1], p[0]))  # OK
```

**To Reproduce**
Run on that code block and it produces an error message of,

```
/typing_experiments/callable_bidirectional1.py
  /typing_experiments/callable_bidirectional1.py:14:52 - error: "__getitem__" method not defined on type "object*" (reportGeneralTypeIssues)
  /typing_experiments/callable_bidirectional1.py:14:58 - error: "__getitem__" method not defined on type "object*" (reportGeneralTypeIssues)
```

**Expected behavior**
I'd expect both x and y to be good. y did work for a couple versions and looks to regressed from 1.1.307 -> 1.1.308.

**VS Code extension or command-line**
pyright cli. The code type checked from 1.1.305 -> 1.1.307 and started producing an error from 1.1.308.

**Additional context**
The Paramspec is necessary even though it's empty here. Simpler version that uses only TypeVar type checks fine. Also in real version of this the x: Callback[tuple[int, int]]  is a bit more implicit and happens like,

```python
def make_foo() -> Callback[tuple[int, int]]:
  return Callback(lambda p: (p[1], p[0]))
```

The work around of y is a little awkward for classes generic in the stub, but not in runtime implementation (doable with TYPE_CHECKING tricks though).

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.

I'm surprised that this ever worked. There was no code in the lambda bidirectional type inference code paths for dealing with ParamSpecs. If it worked (even partially), that was by chance only.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.310, which I just published. It will also be included in a future release of pylance.
